### PR TITLE
Gracefully wait for streams to complete when pipe completes

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/aspnet/KestrelHttpServer/issues/2793.

### CODE REVIEW

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 1]
```diff
@@ -1084,7 +1078,7 @@ public async Task DATA_Received_NoStreamWindowSpace_ConnectionError()
             await SendDataAsync(1, _maxData, endStream: false);
 
             await WaitForConnectionErrorAsync<Http2ConnectionErrorException>(
-                ignoreNonGoAwayFrames: false,
+                ignoreNonGoAwayFrames: true,
```
[the remark this answers is not in the document]
Comment (halter73) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: No headers frame would be written if you switched this test back to using `_waitForAbortApplication`, right?

[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: Yup, my bad.

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 2]
```diff
@@ -891,13 +899,19 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
...
+                        UpdateState(Http2ConnectionState.Closed);
+
+                        // Wake up request processing loop so the connection can complete if there are no pending requests
+                        Input.CancelPendingRead();
+                    }
 
-                    // Wake up request processing loop so the connection can complete if there are no pending requests
-                    Input.CancelPendingRead();
+                    // Complete the task waiting on all streams to finish
+                    _streamsCompleted.TrySetResult(null);
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: What if the state is still Open?

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs [hunk 3]
```diff
@@ -266,6 +272,8 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                         stream.Abort(connectionError);
                     }
 
+                    await _streamsCompleted.Task;
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Timeout/cancellationToken?

File: test/Kestrel.Core.Tests/Http2ConnectionTests.cs [hunk 4]
```diff
@@ -1065,12 +1065,6 @@ public async Task DATA_Received_StreamClosedImplicitly_ConnectionError()
         [Fact]
         public async Task DATA_Received_NoStreamWindowSpace_ConnectionError()
         {
-            // I hate doing this, but it avoids exceptions from MemoryPool.Dipose() in debug mode. The problem is since
```
Comment (Tratcher) [inline test/Kestrel.Core.Tests/Http2ConnectionTests.cs]: There are more of these in Http2StreamTests https://github.com/aspnet/KestrelHttpServer/blob/bd3a8a3511238a87c61d9ed798db373c1b88dd78/test/Kestrel.Core.Tests/Http2StreamTests.cs#L967

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 2 above)
Comment (JunTaoLuo) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: true

File: src/Kestrel.Core/Internal/Http2/Http2Connection.cs (hunk 3 above)
Comment (JunTaoLuo) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Is the abortive shutdown path via Abort(...) not sufficient?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: That's a good start, but right now you have no guarantee that the Task here will ever complete.

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: I don't think we need to have a timeout or a CancellationToken here. What we have now is in line with the behavior for HTTP/1. If a connection's ProcessRequestsAsync method doesn't complete before the shutdown timeout CancellationToken fires, then CloseAllConnectionsAsync will complete and AbortAllConnectionsAsync will be called.

It's think it's a good idea to complete this TCS in Http2Connection.Abort() though.

Comment (JunTaoLuo) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Hmm looks like the only impact if the task doesn't complete is if there are outstanding streams. After the graceful timeout, the abortive path takes over and still completes/aborts the _frameWriter. The only difference is that Input.Complete() is not called. Is that the concern?

Comment (halter73) [reply src/Kestrel.Core/Internal/Http2/Http2Connection.cs]: Now that I think about it more, I think it's a bad idea to complete this TCS in Abort. The socket get's closed in abort (if it isn't already), and that itself might unblock the application. AbortAllConnectionsAsync will timeout after a second if that's not the case anyway.

**Pull request merged**